File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    hash = {}
    hap = {}
    for i in range(len(genres)):
        if genres[i] not in hash:
            hash[genres[i]] = [[i, plays[i]], [-1,-1]]
            hap[genres[i]] = plays[i]

        #     첫번째보다 크면
        elif hash[genres[i]][0][1] < plays[i]:
            hash[genres[i]] = [[i, plays[i]],hash[genres[i]][0]]
            hap[genres[i]] += plays[i]

        #     첫번째랑 같으면
        elif hash[genres[i]][0][1] == plays[i]:
            hash[genres[i]] = [min(hash[genres[i]][0], [i, plays[i]]), max(hash[genres[i]][0], [i, plays[i]])]
            hap[genres[i]] += plays[i]
        #     두번째보다 크면
        elif hash[genres[i]][1][1] < plays[i]:
            hash[genres[i]] = [hash[genres[i]][0], [i, plays[i]]]
            hap[genres[i]] += plays[i]
        else:
            hap[genres[i]] += plays[i]

    sorted_hap = sorted(hap.items(), key = lambda item: item[1], reverse = True)
    answer = []
    for s in sorted_hap:
        for h in hash[s[0]]:
            if h[1] != -1:
                answer.append(h[0])
    return answer

File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.java
import java.util.HashMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

class Solution {
    public ArrayList<Integer> solution(String[] genres, int[] plays) {
        ArrayList<Integer> answer = new ArrayList<Integer>();
        HashMap<String, ArrayList<Integer>> map = new HashMap<String, ArrayList<Integer>>();
        HashMap<String, Integer> hap = new HashMap<String, Integer>();
        
        for (int i = 0; i < genres.length; i++) {
            ArrayList<Integer> value = map.get(genres[i]);
            if (value == null) value = new ArrayList<Integer>();
            value.add(i);
            map.put(genres[i], value);
            
            Integer s = hap.get(genres[i]);
            if (s == null) s = 0;
            s += plays[i];
            hap.put(genres[i], s);
        }
        List<Map.Entry<String, Integer>> entryList = new ArrayList<>(hap.entrySet());
        entryList.sort((e1, e2) -> -e1.getValue().compareTo(e2.getValue()));

        int idx1, idx2, max1, max2;
        for (Map.Entry<String, Integer> entry : entryList) {
            
            ArrayList<Integer> indexList = map.get(entry.getKey());
            idx1 = idx2 = max1 = max2 = 0;
            for (int i = 0; i < indexList.size(); i++) {
                if (plays[indexList.get(i)] > max1) {
                    max2 = max1;
                    idx2 = idx1;
                    max1 = plays[indexList.get(i)];
                    idx1 = indexList.get(i);
                } else if (plays[indexList.get(i)] > max2) {
                    max2 = plays[indexList.get(i)];
                    idx2 = indexList.get(i);
                }
            }
            if (idx1 != -1) answer.add(idx1);
            if (idx2 != -1) answer.add(idx2);
        }        
        
        return answer;
    }
}